lib/http_aws_sigv4.c

```diff
@@ -66,6 +70,198 @@ static void sha256_to_hex(char *dst, unsigned char *sha, size_t dst_l)
+    head = NULL;
+  }
+#ifdef DEBUGBUILD
+  else if(force_host) {
+    head = curl_slist_append(NULL, "host:127.0.0.1");
+    if(!head)
+      goto fail;
+  }
+#endif
+  else {
+    char *full_host;
+
+    if(data->state.aptr.host) {
+      size_t pos;
+
+      full_host = strdup(data->state.aptr.host);
+      /* remove /r/n as the separator for canonical request must be '\n'
+       * and are readd after */
+      pos = strlen(full_host);
+      if(pos > 1 && full_host[pos - 1] == '\n') {
+        full_host[pos - 1] = 0;
+        if(pos > 2 && full_host[pos - 2] == '\r')
+          full_host[pos - 2] = 0;
+      }
+    }
+    else {
+      full_host = curl_maprintf("host:%s", hostname);
+    }
+
+    if(!full_host)
+      goto fail;
+    head = Curl_slist_append_nodup(NULL, full_host);
+    if(!head)
+      goto fail;
+  }
+
+
+  if(!date_hdr_key || !date_full_hdr)
+    goto fail;
+
+  for(l = data->set.headers; l; l = l->next) {
+    tmp_head = curl_slist_append(head, l->data);
+    if(!tmp_head)
+      goto fail;
+    head = tmp_head;
+  }
+
+  /* remove whitespace, and lowercase each headers */
+  for(l = head; l; l = l->next) {
+    char *tmp = l->data, *tmp2;
+    size_t pos;
+
+    for(pos = 0; tmp[pos] && tmp[pos] != ':'; ++pos);
+    Curl_strntolower(l->data, l->data, pos);
+
+    tmp += pos;
+    if(!*tmp)
+      continue;
+    ++tmp;
+    tmp2 = tmp;
+    while(*tmp2 == ' ' || *tmp2 == '\t') {
+      ++tmp2;
+    }
+    if(tmp != tmp2)
+      memmove(tmp, tmp2, strlen(tmp2) + 1);
+
+    while(*tmp) {
+      tmp2 = tmp;
+      while(*tmp2 == ' ' || *tmp2 == '\t') {
+        ++tmp2;
+      }
+
+      if(tmp != tmp2) {
+        if(*tmp2 == 0)
+          /* remove spaces at the end */
+          *tmp = 0;
+        else {
+          *tmp++ = ' '; /* if tab convert to space */
+          if (*tmp == ' ' || *tmp == '\t')
+            memmove(tmp, tmp2, strlen(tmp2));
+        }
+        tmp = tmp2;
+      }
+      else {
+        ++tmp;
+      }
+    }
+  }
+
+  *date_header = find_date_hdr(data, date_hdr_key);
+  if(!*date_header) {
+    tmp_head = curl_slist_append(head, date_full_hdr);
+    if(!tmp_head)
+      goto fail;
+    tmp_head = head;
+    *date_header = curl_maprintf("%s: %s", date_hdr_key, timestamp);
+  }
+  else {
+    char *tmp;
+
+    *date_header = strdup(*date_header);
+    tmp = strchr(*date_header, ':');
+    if(!tmp)
+      goto fail;
+    ++tmp;
+    while(*tmp == ' ' || *tmp == '\t')
+      ++tmp;
+    strncpy(timestamp, tmp, TIMESTAMP_SIZE);
+    timestamp[TIMESTAMP_SIZE - 1] = 0;
+  }
+
+  /* reorder */
+  for(again = 1; again;) {
+    again = 0;
+    for(l = head; l; l = l->next) {
+      struct curl_slist *next = l->next;
+
+      if(next && strcmp(l->data, next->data) > 0) {
+        char *tmp = l->data;
+
+        l->data = next->data;
+        next->data = tmp;
+        again = 1;
+      }
+    }
+  }
+
+  for(l = head; l; l = l->next) {
+    char *tmp = *canonical_headers;
```

Comment:
note, that I'm presently refactoring this code to use dynbuf instead of curl_maprintf for this chunk.